src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_Checksum.cs

```diff
@@ -20,6 +20,116 @@
+                if (!_lazyProjectChecksums.TryGetValue(projectId, out checksums) ||
+                    checksums == null)
+                {
+                    stateChecksums = null;
+                    return false;
+                }
+            }
+
+            return checksums.TryGetValue(out stateChecksums);
+        }
+
+        public Task<SolutionCompilationStateChecksums> GetStateChecksumsAsync(CancellationToken cancellationToken)
+            => _lazyChecksums.GetValueAsync(cancellationToken);
+
+        public async Task<Checksum> GetChecksumAsync(CancellationToken cancellationToken)
+        {
+            var collection = await GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+            return collection.Checksum;
+        }
+
+        /// <summary>Gets the checksum for only the requested project (and any project it depends on)</summary>
+        public async Task<SolutionCompilationStateChecksums> GetStateChecksumsAsync(
+            ProjectId projectId,
+            CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(projectId);
+
+            AsyncLazy<SolutionCompilationStateChecksums>? checksums;
+            lock (_lazyProjectChecksums)
+            {
+                if (!_lazyProjectChecksums.TryGetValue(projectId, out checksums))
+                {
+                    checksums = Compute(projectId);
+                    _lazyProjectChecksums.Add(projectId, checksums);
+                }
+            }
+
+            var collection = await checksums.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            return collection;
+
+            // Extracted as a local function to prevent delegate allocations when not needed.
+            AsyncLazy<SolutionCompilationStateChecksums> Compute(ProjectId projectId)
+                => AsyncLazy.Create(c => ComputeChecksumsAsync(projectId, c));
+        }
+
+        /// <summary>Gets the checksum for only the requested project (and any project it depends on)</summary>
+        public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationToken cancellationToken)
+        {
+            var checksums = await GetStateChecksumsAsync(projectId, cancellationToken).ConfigureAwait(false);
+            return checksums.Checksum;
+        }
+
+        private async Task<SolutionCompilationStateChecksums> ComputeChecksumsAsync(
+            ProjectId? projectId,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                using (Logger.LogBlock(FunctionId.SolutionCompilationState_ComputeChecksumsAsync, this.SolutionState.FilePath, cancellationToken))
+                {
+                    var solutionStateChecksum = projectId == null
+                        ? await this.SolutionState.GetChecksumAsync(cancellationToken).ConfigureAwait(false)
+                        : await this.SolutionState.GetChecksumAsync(projectId, cancellationToken).ConfigureAwait(false);
```

Comment:
we pass the desired top level project (or nothing) into the SolutionState, which computes the same way as before, just without the frozenSG stuff.